File: lib/models/remote_config.dart
/// Mirrors the backend's GET /config response — the update switchboard.
class RemoteConfig {
  final int latestVersionCode;
  final String latestVersionName;
  final int forceUpdateBelow;
  final List<String> changelog;
  final String? announcement;
  final Map<String, bool> features;

  /// Self-hosted update channel (sideload builds): direct APK download URL
  /// + its sha256, published via the backend's POST /admin/apk. Null when
  /// no build has been uploaded — the app then falls back to store flows.
  final String? apkUrl;
  final String? apkSha256;

  const RemoteConfig({
    this.latestVersionCode = 1,
    this.latestVersionName = '0.1.0',
    this.forceUpdateBelow = 0,
    this.changelog = const [],
    this.announcement,
    this.features = const {},
    this.apkUrl,
    this.apkSha256,
  });

  factory RemoteConfig.fromJson(Map<String, dynamic> j) => RemoteConfig(
        latestVersionCode: j['latestVersionCode'] ?? 1,
        latestVersionName: j['latestVersionName'] ?? '0.1.0',
        forceUpdateBelow: j['forceUpdateBelow'] ?? 0,
        changelog: List<String>.from(j['changelog'] ?? const []),
        announcement: j['announcement'],
        features: Map<String, bool>.from(j['features'] ?? const {}),
        apkUrl: j['apkUrl'],
        apkSha256: j['apkSha256'],
      );

  bool isEnabled(String feature) => features[feature] == true;
}

File: app/src/main/java/com/myassistant/app/update/RemoteConfig.kt
package com.myassistant.app.update

import com.myassistant.app.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request

/**
 * SERVER-DRIVEN FEATURES — how "updates without rebuilding" works.
 *
 * The app fetches this config from YOUR backend on every launch.
 * When you ship a new AI capability on the server (a new tool, a new
 * language, a new card type), you flip its flag here and every
 * installed app picks it up instantly — no Play Store release needed.
 *
 * What CAN be updated this way (Play-policy safe):
 *   - AI behaviour, prompts, models, new assistant capabilities (all server-side)
 *   - Feature flags: show/hide screens and buttons already in the app
 *   - Announcements, changelog text, quality settings
 *
 * What CANNOT (requires a Play Store release):
 *   - New native code / screens that don't exist in the APK yet.
 *     Downloading and executing code outside Play violates Google Play
 *     policy and will get the app removed. For those, UpdateManager
 *     triggers the official Play in-app update flow instead.
 */
@Serializable
data class RemoteConfig(
    val latestVersionCode: Int = 1,
    val latestVersionName: String = "0.1.0",
    val forceUpdateBelow: Int = 0,          // versions below this must update
    val changelog: List<String> = emptyList(),
    val announcement: String? = null,        // e.g. "New: Malayalam voice is live!"
    val features: Map<String, Boolean> = emptyMap() // e.g. "voice_mode" to true
)

object ConfigRepository {
    private val client = OkHttpClient()
    private val json = Json { ignoreUnknownKeys = true }

    @Volatile var current: RemoteConfig = RemoteConfig()
        private set

    suspend fun refresh(): RemoteConfig = withContext(Dispatchers.IO) {
        try {
            val req = Request.Builder()
                .url("${BuildConfig.BASE_URL}/config?v=${BuildConfig.VERSION_CODE}")
                .build()
            client.newCall(req).execute().use { resp ->
                if (resp.isSuccessful) {
                    current = json.decodeFromString<RemoteConfig>(resp.body!!.string())
                }
            }
        } catch (_: Exception) {
            // Offline or server down — keep last known config. Never crash on config.
        }
        current
    }

    fun isEnabled(feature: String): Boolean = current.features[feature] == true
}
